core/src/main/scala/org/apache/spark/TaskEndReason.scala

```diff
@@ -97,13 +103,26 @@ case class ExceptionFailure(
     description: String,
     stackTrace: Array[StackTraceElement],
     fullStackTrace: String,
-    metrics: Option[TaskMetrics])
+    metrics: Option[TaskMetrics],
+    exceptionWrapper: Option[ThrowableSerializationWrapper])
```

Comment:
Agreed - there's a test I added in FailureSuite that uses NonDeserializableUserException to check this exact case.